### PR TITLE
Provide option to implement `shouldUpgrade` on webSockets

### PR DESCRIPTION
**Pull request by code28**
Adds support for a custom implementation of `shouldUpgrade` for websockets. This can be used to switch to specific subprotocols or to deny upgrading by returning `nil`.

```swift
routes.webSocket("authedecho", shouldUpgrade: { req in
    guard req.auth.has(User.self) else {
        return req.eventLoop.future(nil)
    }
    return req.eventLoop.future([:])            
}, onUpgrade: { req, ws in
    print(ws)
})
```

### CODE REVIEW

**Comment by ChaosCoder** [comment]
This PR looks good and is very much needed. 

There was a hacky solution in Vapor 3 for this kind of authorization on `shouldUpgrade` as described in this blog post: https://blog.kaltoun.cz/websockets-and-authentication-in-vapor3/ There were problems with the implementation in Vapor 3 though as the code had to be synchronous (no futures), which is quite an exception in Vapor.

In Vapor 4 the hacky solution does not work *at all*. The underlying asynchronous  callback by NIO2 is provided, but it is not exposed and accessible from the outside in Vapor 4. This PR seems to fix the problem. :+1:

File: Sources/Vapor/Routing/RoutesBuilder+WebSocket.swift [hunk 1]
```diff
@@ -15,20 +15,23 @@ extension RoutesBuilder {
     public func webSocket(
         _ path: PathComponent...,
         maxFrameSize: WebSocketMaxFrameSize = .`default`,
+        shouldUpgrade: ((Request) -> EventLoopFuture<HTTPHeaders?>)? = nil,
```
Comment (tanner0101) [inline Sources/Vapor/Routing/RoutesBuilder+WebSocket.swift:18]: The double-optional here is a bit hard to parse. What about putting the default here?
```swift
shouldUpgrade: ((Request) -> EventLoopFuture<HTTPHeaders?>) = { 
    $0.eventLoop.makeSucceededFuture([:]) 
},
```

File: Sources/Vapor/Routing/RoutesBuilder+WebSocket.swift [hunk 2]
```diff
@@ -15,20 +15,23 @@ extension RoutesBuilder {
...
     public func webSocket(
         _ path: [PathComponent],
         maxFrameSize: WebSocketMaxFrameSize = .`default`,
+        shouldUpgrade: ((Request) -> EventLoopFuture<HTTPHeaders?>)? = nil,
         onUpgrade: @escaping (Request, WebSocket) -> ()
     ) -> Route {
         return self.on(.GET, path) { request -> Response in
             let res = Response(status: .switchingProtocols)
-            res.upgrader = .webSocket(maxFrameSize: maxFrameSize, onUpgrade: { ws in
+            let shouldUpgradeWrapped = shouldUpgrade.map { shouldUpgrade in { shouldUpgrade(request) } }
```
Comment (tanner0101) [inline Sources/Vapor/Routing/RoutesBuilder+WebSocket.swift:33]: Some newlines here would help readability.

**Review by tanner0101 — APPROVED** [review]
Thanks for your patience. This looks great! Just a minor API nit but otherwise this is good to merge.

Would you mind formatting the PR body as release notes? See channel.eventLoop.makeSucceededFuture([:])

Thanks!

File: Sources/Vapor/Routing/RoutesBuilder+WebSocket.swift (hunk 1 above)
Comment (code28) [reply Sources/Vapor/Routing/RoutesBuilder+WebSocket.swift]: I was unsure about this, because the default has to be put at two places. But since this leads to a lot better readability, I think it's worth it.

**Comment by code28** [comment]
Thank you for your review!
I applied your feedback and edited the PR body. I hope, it's fine for you.

**Comment by siemensikkema** [comment]
@code28 I know it's approved already but I think it'd be great if we could have some API docs on the public functions. I know, they were not there before but with this added functionality, the only way to know what the parameter does is when you're familiar with the underlying implementation. Also a test to prove that functionality works (to prove it's connected correctly, not to test the NIOWebSocket itself of course) would be awesome and function as more documentation in a way.

**Comment by code28** [comment]
@siemensikkema Thanks for your reply! Sure, I'm happy to add them, if that'll lead to this getting merged. :) Can you point me to an example, how (and in which style/to which extend) to add the API docs?

**Comment by siemensikkema** [comment]
@code28 I'd like to align the API doc style across all of Vapor and aim for 100% coverage but for now I think Xcode's default style would do. Something like:

```
    /// <#Description#>
    /// - Parameters:
    ///   - path: <#path description#>
    ///   - maxFrameSize: <#maxFrameSize description#>
    ///   - shouldUpgrade: <#shouldUpgrade description#>
    ///   - onUpgrade: <#onUpgrade description#>
    /// - Returns: <#description#>
```

See also for example the API docs in `ContentContainer.swift`.

**Comment by code28** [comment]
@siemensikkema Thanks for the hint. I added those API docs (and rebased on current `master` branch.) I took the description for `shouldUpgrade` and `maxFrameSize` from NIO ([those comments](https://github.com/apple/swift-nio/blob/366850acddd7dccf9d29e6dcbb143199a217a271/Sources/NIOWebSocket/NIOWebSocketServerUpgrader.swift#L104-L115)). This is probably not okay with the license, but I thought, the description is the most precise. Maybe it's better to just add a reference? What do you think?

**Comment by siemensikkema** [comment]
@code28 I think we can do with a less precise and slightly rephrased single sentence description for each to be safe and then a reference to NIO as you suggested. Did you also have a change to look at implementing a test?

**Comment by code28** [comment]
@siemensikkema I rephrased the docs and added a test to check, if the `shouldUpgrade` callback gets called. Is this as you imagined?

File: Tests/VaporTests/RouteTests.swift [hunk 3]
```diff
@@ -351,4 +351,25 @@ final class RouteTests: XCTestCase {
...
+    
+    func testWebsocketUpgrade() throws {
+        let app = Application(.testing)
+        defer { app.shutdown() }
+        
+        let testMarkerHeaderKey = "TestMarker"
+        let testMarkerHeaderValue = "addedInShouldUpgrade"
+        
+        app.routes.webSocket("customshouldupgrade", shouldUpgrade: { req in
+            return req.eventLoop.future([testMarkerHeaderKey:testMarkerHeaderValue])
```
Comment (siemensikkema) [inline Tests/VaporTests/RouteTests.swift:363]:
```suggestion
            return req.eventLoop.future([testMarkerHeaderKey: testMarkerHeaderValue])
```

Comment (siemensikkema) [reply Tests/VaporTests/RouteTests.swift:363]: @code28 One last thing and this is good to go 🙂

**Comment by code28** [comment]
Done :)

**Review by siemensikkema — APPROVED** [review]
Looks great! Thanks for going the extra mile with API docs and tests

**Comment by tanner0101** [comment]
These changes are now available in [4.33.0](https://github.com/vapor/vapor/releases/tag/4.33.0)

**Comment by klaas** [comment]
What should happen if `shouldUpgrade` returns a `nil` future?

In `NIOWebSocketServerUpgrader.swift` a `NIOWebSocketUpgradeError.unsupportedWebSocketTarget` is thrown and the `onUpgrade` callback will not be called.

But in my WebSocket client nothing is received until the connection times out 🤔?

Using https://github.com/vapor/vapor/pull/2498 works fine and I can manually return a response failure before upgrading the WebSocket.

**Comment by code28** [comment]
@klaas I guess this depends on your websocket implementation. I'm not 100% sure, what the actual HTTP reaction is, but it's supposed to deny upgrading (as mentioned in NIO [here](https://github.com/apple/swift-nio/blob/366850acddd7dccf9d29e6dcbb143199a217a271/Sources/NIOWebSocket/NIOWebSocketServerUpgrader.swift#L104-L115)).
So, not calling `onUpgrade`, when `shouldUpgrade` returns `nil`, is exactly the expected behaviour. The upgrade is denied, so no `onUpgrade`.

**Comment by klaas** [comment]
@code28 Thanks for you reply.

The WebSocket [RFC 6455](https://datatracker.ietf.org/doc/html/rfc6455) states: "Any status code other than 101 indicates that the WebSocket handshake has not completed and that the semantics of HTTP still apply".

And `automaticErrorHandling` is set to `false` in https://github.com/vapor/vapor/blob/73d57959a96ceca719558e70c320b2f729d2b25b/Sources/Vapor/HTTP/Server/HTTPServerUpgradeHandler.swift#L55-L59 


How am I supposed to "handle my own errors" as stated in the api doc that you supplied? Any hint would be appreciated 🙂.

Right now I do the manual upgrade (as in https://github.com/vapor/vapor/pull/2498 ) and return a HTTP 400 or 401. My client is using Foundations `URLSessionWebSocketTask` and gets a `didCompleteWithError` delegate call. It works fine for us at the moment, but it would be nice to understand how to use the regular ".webSocket" route in Vapor.
